vm/src/protocol/mapping.rs

```diff
@@ -1,126 +1,167 @@
+        PyDict,
     },
-    function::IntoPyObject,
-    IdProtocol, PyObject, PyObjectRef, PyObjectWrap, PyResult, TryFromObject, TypeProtocol,
+    common::lock::OnceCell,
+    function::IntoPyResult,
+    IdProtocol, PyObject, PyObjectPayload, PyObjectRef, PyObjectView, PyResult, TypeProtocol,
     VirtualMachine,
 };
-use std::borrow::Borrow;
 
 // Mapping protocol
 // https://docs.python.org/3/c-api/mapping.html
 #[allow(clippy::type_complexity)]
-#[derive(Default)]
+#[derive(Default, Copy, Clone)]
 pub struct PyMappingMethods {
-    pub length: Option<fn(PyObjectRef, &VirtualMachine) -> PyResult<usize>>,
-    pub subscript: Option<fn(PyObjectRef, PyObjectRef, &VirtualMachine) -> PyResult>,
+    pub length: Option<fn(&PyMapping, &VirtualMachine) -> PyResult<usize>>,
+    pub subscript: Option<fn(&PyMapping, &PyObject, &VirtualMachine) -> PyResult>,
     pub ass_subscript:
-        Option<fn(PyObjectRef, PyObjectRef, Option<PyObjectRef>, &VirtualMachine) -> PyResult<()>>,
+        Option<fn(&PyMapping, &PyObject, Option<PyObjectRef>, &VirtualMachine) -> PyResult<()>>,
 }
 
-#[derive(Debug, Clone)]
-#[repr(transparent)]
-pub struct PyMapping<T = PyObjectRef>(T)
-where
-    T: Borrow<PyObject>;
-
-impl PyMapping<PyObjectRef> {
-    pub fn check(obj: &PyObject, vm: &VirtualMachine) -> bool {
-        obj.class()
-            .mro_find_map(|x| x.slots.as_mapping.load())
-            .map(|f| f(obj, vm).subscript.is_some())
-            .unwrap_or(false)
+#[derive(Clone)]
+pub struct PyMapping<'a> {
+    pub obj: &'a PyObject,
+    methods: OnceCell<PyMappingMethods>,
+}
+
+impl<'a> From<&'a PyObject> for PyMapping<'a> {
+    fn from(obj: &'a PyObject) -> Self {
+        Self {
+            obj,
+            methods: OnceCell::new(),
+        }
     }
+}
 
-    pub fn methods(&self, vm: &VirtualMachine) -> PyMappingMethods {
-        let obj_cls = self.0.class();
-        for cls in obj_cls.iter_mro() {
-            if let Some(f) = cls.slots.as_mapping.load() {
-                return f(&self.0, vm);
-            }
+impl AsRef<PyObject> for PyMapping<'_> {
+    fn as_ref(&self) -> &PyObject {
+        self.obj
+    }
+}
+
+impl<'a> PyMapping<'a> {
+    pub fn with_methods(obj: &'a PyObject, methods: PyMappingMethods) -> Self {
+        Self {
+            obj,
+            methods: OnceCell::from(methods),
+        }
+    }
+
+    pub fn try_protocol(obj: &'a PyObject, vm: &VirtualMachine) -> PyResult<Self> {
+        let zelf = Self::from(obj);
+        if zelf.has_protocol(vm) {
+            Ok(zelf)
+        } else {
+            Err(vm.new_type_error(format!("{} is not a mapping object", zelf.obj.class())))
         }
-        PyMappingMethods::default()
     }
 }
 
-impl<T> PyMapping<T>
-where
-    T: Borrow<PyObject>,
-{
-    pub fn new(obj: T) -> Self {
-        Self(obj)
+impl PyMapping<'_> {
+    // PyMapping::Check
+    pub fn has_protocol(&self, vm: &VirtualMachine) -> bool {
+        self.methods(vm).subscript.is_some()
+    }
```

Comment:
How about just using `check`? PyIter also adopt `check` for `PyIter::Check`

Reply:
make sense, but I think I'd like to keep fn try_protocol() as it is more rust style.